2018 Robot/src/org/usfirst/frc/team4999/robot/MoPrefs.java

```diff
@@ -0,0 +1,76 @@
+package org.usfirst.frc.team4999.robot;
+
+import edu.wpi.first.wpilibj.Preferences;
+
+public class MoPrefs {
+	
+	private Preferences prefs = Preferences.getInstance();
+	private static MoPrefs instance;
+	
+	private static final double MOVE_P = 0.2;
+	private static final double MOVE_I = 0;
+	private static final double MOVE_D = 0.05;
+	private static final double TURN_P = 0.2;
+	private static final double TURN_I = 0;
+	private static final double TURN_D = 0.05;
+	private static final int 	MAX_PI_DELAY = 200;
+	private static final double AUTO_SPEED = 0.25;
+	
+	
+	
+	private void checkDouble(String key, double def) {
+		if(!prefs.containsKey(key)) {
+			prefs.putDouble(key, def);
+		}
+	}
+	
+	private void checkInt(String key, int def) {
+		if(!prefs.containsKey(key)) {
+			prefs.putInt(key, def);
+		}
+	}
+	
+	public static MoPrefs getInstance() {
+		if(instance == null) {
+			instance = new MoPrefs();
+		}
+		return instance;
+	}
+	
+	private MoPrefs() {
+		checkDouble("MOVE_P", MOVE_P);
+		checkDouble("MOVE_I", MOVE_I);
+		checkDouble("MOVE_D", MOVE_D);
+		checkDouble("TURN_P", TURN_P);
+		checkDouble("TURN_I", TURN_I);
+		checkDouble("TURN_D", TURN_D);
+		checkInt("MAX_PI_DELAY", MAX_PI_DELAY);
+		checkDouble("AUTO_SPEED", AUTO_SPEED);
+	}
+	
+	public double getMoveP() {
```

Comment:
curious? Where are the setters?

Reply:
The MoPrefs wraps the [robot preferences](http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/Preferences.html) so that the keys and default values aren't magic values. The constants at the top of the file are default values, and are only used if a key doesn't exist in the robot preferences yet. The values in the preferences can be set via a widget on the SmartDashboard, so the setters are on the SmartDashboard.